src/ring_theory/dedekind_domain/S_integer.lean

```diff
@@ -0,0 +1,116 @@
+can be specialised to the case of a number field or a function field separately.
+
+TODO:
+ * notation for `S`-integers and `S`-units.
+ * proof that `S`-units is the kernel of a map to a product.
+ * proof that `∅`-integers is the usual ring of integers.
+ * finite generation of `S`-units and Dirichlet's `S`-unit theorem.
+
+## Main definitions
+
+ * `set.integer`: `S`-integers.
+ * `set.unit`: `S`-units.
+
+## Main statements
+
+ * `set.units_integer_equiv_unit`: units of `S`-integers are `S`-units.
+
+## References
+
+ * [D Marcus, *Number Fields*][marcus1977number]
+ * [J W S Cassels, A Frölich, *Algebraic Number Theory*][cassels1967algebraic]
+ * [J Neukirch, *Algebraic Number Theory*][Neukirch1992]
+
+## Tags
+
+S integer, S-integer, S unit, S-unit
+-/
+
+open is_dedekind_domain
+
+open_locale non_zero_divisors
+
+namespace set
+
+noncomputable theory
+
+universes u v
+
+variables {R : Type u} [comm_ring R] [is_domain R] [is_dedekind_domain R]
+  (S : set $ height_one_spectrum R) (K : Type v) [field K] [algebra R K] [is_fraction_ring R K]
+
+/-! ## `S`-integers -/
+
+/-- The subring of `S`-integers of `K`. -/
+@[simps] def integer : subring K :=
+{ carrier   := {x : K | ∀ v ∉ S, (v : height_one_spectrum R).valuation x ≤ 1},
+  mul_mem'  := λ _ _ hx hy v hv, by simp only [map_mul, mul_le_one₀ (hx v hv) (hy v hv)],
+  one_mem'  := λ _ _, by simp only [map_one, le_refl],
+  add_mem'  := λ _ _ hx hy v hv, v.valuation.map_add_le (hx v hv) (hy v hv),
+  zero_mem' := λ _ _, by simp only [map_zero, zero_le_one],
+  neg_mem'  := λ _ hx v hv, by simp only [valuation.map_neg, hx v hv] }
+
+lemma integer_eq :
+  S.integer K = ⨅ v ∉ S, (v : height_one_spectrum R).valuation.valuation_subring.to_subring :=
+subring.ext $ λ _, by simpa only [subring.mem_infi]
+
+/-- The subring of `S`-integers of `K` is an algebra over `R`. -/
+instance : algebra R (S.integer K) :=
+{ smul      := λ x y, ⟨algebra_map R K x, λ v _, v.valuation_le_one x⟩ * y,
+  to_fun    := λ x, ⟨algebra_map R K x, λ v _, v.valuation_le_one x⟩,
+  map_one'  := by simpa only [map_one],
+  map_mul'  := λ _ _, by simpa only [map_mul],
+  map_zero' := by simpa only [map_zero],
+  map_add'  := λ _ _, by simpa only [map_add],
+  commutes' := λ _, mul_comm _,
+  smul_def' := λ _ _, rfl }
+
+/-! ## `S`-units -/
+
+/-- The subgroup of `S`-units of `Kˣ`. -/
+@[simps] def unit : subgroup Kˣ :=
+{ carrier  := {x : Kˣ | ∀ v ∉ S, (v : height_one_spectrum R).valuation (x : K) = 1},
+  mul_mem' := λ _ _ hx hy v hv, by rw [units.coe_mul, map_mul, hx v hv, hy v hv, one_mul],
+  one_mem' := λ _ _, map_one _,
+  inv_mem' := λ _ hx v hv, by rw [map_units_inv, hx v hv, inv_one] }
+
+lemma unit_eq :
+  S.unit K = ⨅ v ∉ S, (v : height_one_spectrum R).valuation.valuation_subring.unit_group :=
+begin
+  ext,
+  simp only [subgroup.mem_infi],
+  exact forall₂_congr (λ _ _, (valuation.is_equiv_iff_val_eq_one _ _).mp $
+                        valuation.is_equiv_valuation_valuation_subring _)
```

Comment:
I suggest that you extract
```
lemma _root_.valuation.mem_unit_group_iff {Γ} [linear_ordered_comm_group_with_zero Γ]
  (v : _root_.valuation K Γ) {x : Kˣ} : x ∈ v.valuation_subring.unit_group ↔ v x = 1 :=
(valuation.is_equiv_iff_val_eq_one _ _).mp (valuation.is_equiv_valuation_valuation_subring _).symm
```
and put it under [valuation_subring.mem_unit_group_iff](https://leanprover-community.github.io/mathlib_docs/ring_theory/valuation/valuation_subring.html#valuation_subring.mem_unit_group_iff).